### PR TITLE
feat: Update form metadata API dev complete [401088]

### PR DESCRIPTION
**Pull request by KiranVarma02**
https://dev.azure.com/defragovuk/DEFRA-CDP/_workitems/edit/401088

form metadata updatedAt and updatedBy are not getting updated as part of this PR but should be done once we get clarify from BA on how to show it in UI

### CODE REVIEW

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:211]: We only want to update the slug if  `formUpdate.title` is provided, this will set the updated slug to `undefined` if title is missing in the update request.

Comment (KiranVarma02) [reply src/api/forms/service.js:211]: I thought about this but didn't fix it yet. I will fix it during my testing 👍

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:199]: `(draft)` can be removed from here, since these changes do not apply to the draft and live.

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:235]: `(draft)` can be removed from here, since these changes do not apply to the draft and live.

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:225]: this line should be before L223, otherwise it's updated and not updating (which is already covered at the end of the function.

File: src/routes/forms.js
Comment (alexluckett) [inline src/routes/forms.js:103]: worth a comment here stating why you're doing this

e.g.

"Take the form metadata update schema and make all fields optional. This acts similar to `Partial<T>` in Typescript."

File: src/routes/forms.js
Comment (alexluckett) [inline src/routes/forms.js:92]: this log is already handled inside the service call, you can remove it here

**Comment by alexluckett** [comment]
Also check out your checks:

```
src/api/forms/service.test.js(212,51): error TS2554: Expected 2 arguments, but got 3.
src/api/forms/service.test.js(222,50): error TS2554: Expected 2 arguments, but got 3.
```

File: src/api/forms/service.js
Comment (davidjamesstone) [inline src/api/forms/service.js:221]: Unneccessary transaction?

File: src/api/forms/service.js
Comment (davidjamesstone) [inline src/api/forms/service.js:211]: We're creating a slug even though `formUpdate.title` might not be present.
How about:

```
const updatedForm = { ...formUpdate }

if (formUpdate.title) {
  updatedForm.slug = slugify(formUpdate.title)
}
```

File: src/api/types.js
Comment (davidjamesstone) [inline src/api/types.js:7]: `{ force: boolean }` minor spacing change to match others `@typedef`s

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js:227]: you say that because its just 1 DB operation? I think I agree! 

I will wait for your reply and remove

File: src/api/forms/service.js
Comment (KiranVarma02) [inline src/api/forms/service.js:223]: I think I don't need this line but update is expecting this arg. I am checking

File: src/routes/forms.js
Comment (colinrotherham) [inline src/routes/forms.js:290]: Just a reminder we can start using the new `@import` syntax now TypeScript 5.5 is installed

```js
/**
 * @import { FormMetadataAuthor, FormMetadataInput } from '@defra/forms-model'
 * @import { AuthCredentials, RequestAuth, ServerRoute, UserCredentials } from '@hapi/hapi'
 * @import { OidcStandardClaims } from 'oidc-client-ts'
 * @import { RequestFormById, RequestFormBySlug, RequestFormDefinition, RequestFormMetadataCreate, RequestFormMetadataUpdateById, RequestRemoveFormById } from '~/src/api/types.js'
 */
```

File: src/api/forms/service.js
Comment (colinrotherham) [inline src/api/forms/service.js:204]: Can I suggest a tweak to this too?

Otherwise you could set `{ title: '' }` (falsy value) and it wouldn't be caught by this check

```suggestion
    if (form.live && 'title' in formUpdate) {
```

File: src/api/types.js
Comment (KiranVarma02) [reply src/api/types.js:8]: when I do Partial here, I get `Property 'id' does not exist on type 'Partial<FormMetadataInput>'.ts(2339)` because I use id from request params

File: src/routes/forms.js
Comment (KiranVarma02) [reply src/routes/forms.js:290]: I will keep that in mind. I think I am going to leave it in this PR and change all in separate PR, if that's okay?

Comment (colinrotherham) [reply src/routes/forms.js:290]: Yeah course, thanks Kiran

File: src/api/types.js
Comment (colinrotherham) [reply src/api/types.js:8]: Ahh that's good, it's found an issue then

i.e. We're supposed to be using the `id` from the URL params not the request payload

**Comment by nayij** [comment]
pls don't forget to update read me

File: src/api/types.js
Comment (KiranVarma02) [reply src/api/types.js:8]: Ignore all above, I am not sure what's happening but I don't see error now when I added Partial. 

Its line 88 in forms.js

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:229]: This has changed quite a bit since we last spoke, what made you change it? We're now returning an updated slug but not setting this in the database.

Your previous approach was almost there.

Taking Dave's comment as inspiration:

```javascript 

const updatedForm = { ...formUpdate }

if (formUpdate.title) {
  updatedForm.slug = slugify(formUpdate.title)
}
```

with something like this, the return would be something like `return updatedForm.slug ?? form.slug`

File: src/api/forms/service.js
Comment (alexluckett) [inline src/api/forms/service.js:241]: since we're not using the session, try dropping this try/catch and drop the `session` parameter

**Review by alexluckett — CHANGES_REQUESTED** [review]

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js:229]: If I remember correct, when I tried above I get 

```
Property 'slug' does not exist on type '{ title?: string | undefined; organisation?: string | undefined; teamName?: string | undefined; teamEmail?: string | undefined; }'
```

after which I should have done this - 

```
 const updatedForm = {
      slug: form.slug,
      ...formUpdate
    }

    if (formUpdate.title) {
      updatedForm.slug = slugify(formUpdate.title)
    }
```

and return updatedForm.slug

this should work I think

File: src/api/forms/repositories/form-metadata-repository.js
Comment (KiranVarma02) [inline src/api/forms/repositories/form-metadata-repository.js:128]: this seems to work to make it optional. Is there any other better way?

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js]: if I spy'd on formMetadata.update and test on args, I would have caught this issue in tests

Comment (colinrotherham) [reply src/api/forms/service.js]: Ah that's because no type was set for `{ ...formUpdate }`

The compiler needs to know in advance what the future (optional) keys might be

```patch
+   /** @satisfies {Partial<FormMetadataDocument>} */
    const updatedForm = {
```

File: src/routes/forms.js
Comment (KiranVarma02) [reply src/routes/forms.js:290]: as agreed, will be addressed in different PR. I added to my todo list

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js]: done

File: src/api/types.js
Comment (KiranVarma02) [reply src/api/types.js]: done

File: README.md
Comment (KiranVarma02) [inline README.md:95]: I will add other endpoints in another PR. I don't want to hold this PR for too long

**Review by davidjamesstone — APPROVED** [review]

**Review by alexluckett — APPROVED** [review]

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js:202]: please check after next commit

File: src/api/forms/service.js
Comment (KiranVarma02) [reply src/api/forms/service.js]: this line doesn't exist. Also, line numbers here doesn't match to current code base. Do you need an old page maybe? please refresh

Comment (davidjamesstone) [reply src/api/forms/service.js]: This is an old outdated comment. You fixed it with this [commit](https://github.com/DEFRA/forms-manager/pull/223/commits/b1954ffe4c4cc31bcd9139932e3b0dcbc26ee80f) yesterday

Comment (KiranVarma02) [reply src/api/forms/service.js]: in my github, I see your comments as 2 days old but they are after your approval. There is section of comments 2 days old which are on top of comments and new section of comments after approval. Something doesn't seem right! but I got your approval, so all good 👍

**Pull request merged**
